design/src/rfcs/rfc0022_refine_builder.md

```diff
@@ -0,0 +1,556 @@
+There is consensus that the new API is an improvement over the pre-existing `OperationRegistryBuilder`/`OperationRegistry`, which is now on its way to deprecation in one of the next releases.  
+
+This RFC builds on top of [RFC 20] to explore an alternative API design prior to its stabilisation.  
+The API proposed in this RFC has been manually implemented for the Pokemon service. You can find the code [here](https://github.com/LukeMathWalker/builder-experiments).
+
+## Overview
+
+Goals:
+- Maximise API ergonomics, with a particular focus on the developer experience for Rust beginners.
+
+Strategy:
+- Reduce type complexity, exposing a less generic API;
+- Provide clearer errors when the service builder is misconfigured.
+
+Trade-offs:
+- Reduce compile-time safety. Missing handlers will be detected at runtime instead of compile-time.
+
+Constraints:
+- There should be no significant degradation in runtime performance (i.e. startup time for applications).
+
+## Handling missing operations
+
+Let's start by reviewing the API proposed in [RFC 20]. We will use the [Pokemon service] as our driving example throughout the RFC.  
+This is what the startup code looks like:
+
+```rust
+#[tokio::main]
+pub async fn main() {
+    // [...]
+    let app = PokemonService::builder()
+        .get_pokemon_species(get_pokemon_species)
+        .get_storage(get_storage)
+        .get_server_statistics(get_server_statistics)
+        .capture_pokemon(capture_pokemon)
+        .do_nothing(do_nothing)
+        .check_health(check_health)
+        .build();
+
+    // Setup shared state and middlewares.
+    let shared_state = Arc::new(State::default());
+    let app = app.layer(&AddExtensionLayer::new(shared_state));
+
+    // Start the [`hyper::Server`].
+    let bind: SocketAddr = /* */;
+    let server = hyper::Server::bind(&bind).serve(app.into_make_service());
+    // [...]
+}
+```
+
+The builder is infallible: we are able to verify at compile-time that all handlers have been provided using the [typestate builder pattern].  
+
+### Compiler errors cannot be tuned
+
+What happens if we stray away from the happy path? We might forget, for example, to add the `check_health` handler.  
+The compiler greets us with this error:
+
+```text
+error[E0277]: the trait bound `MissingOperation: Upgradable<AwsRestJson1, CheckHealth, (), _, IdentityPlugin>` is not satisfied
+  --> pokemon-service/src/bin/pokemon-service.rs:38:10
+   |
+38 |         .build();
+   |          ^^^^^ the trait `Upgradable<AwsRestJson1, CheckHealth, (), _, IdentityPlugin>` is not implemented for `MissingOperation`
+   |
+   = help: the following other types implement trait `Upgradable<Protocol, Operation, Exts, B, Plugin>`:
+             FailOnMissingOperation
+             Operation<S, L>
```

Comment:
I actually think this is one of the nicer error messages you can get. It has enough of the right words `MissingOperation`, `CheckHealth` for it to likely to not cause much confusion.

The error messages you get when passing in a handler that does not satisfy the type signature (not `async`, wrong input / output types) are worse, although those (and perhaps this one too?) can be palliated with a debug macro à la [`axum`'s `debug_handler`](https://docs.rs/axum-macros/latest/axum_macros/attr.debug_handler.html).

I expect this error to not be frequent. Users will likely start off from a template; the new service builder needs to generate Rust docs with an out-of-the-box compilable program, like in https://github.com/awslabs/smithy-rs/pull/1498#issuecomment-1280720925.

Reply:
> The error messages you get when passing in a handler that does not satisfy the type signature (not async, wrong input / output types) are worse, although those (and perhaps this one too?) can be palliated with a debug macro à la [axum's debug_handler](https://docs.rs/axum-macros/latest/axum_macros/attr.debug_handler.html), which I think we direly need anyway.

I don't disagree that those errors are worse, but I don't think that the builder case can be fixed using a debug macro. Or, at least, not in the general case nor easily.
The macro would need to wrap the entire scope where the builder is being used, therefore ruling out refactorings that break the startup logic across multiple functions. We would also have to understand branches in the code we are wrapping to make sure that the all paths lead to a correctly instantiated builder - it gets nasty quite quickly.